tests/end2end/features/conftest.py

```diff
@@ -620,55 +605,104 @@ def check_contents_json(quteproc, text):
+
     active_suffix = ' (active)'
     pinned_suffix = ' (pinned)'
-    tabs = tabs.splitlines()
-    assert len(session['windows']) == 1
-    assert len(session['windows'][0]['tabs']) == len(tabs)
-
-    # If we don't have (active) anywhere, don't check it
-    has_active = any(active_suffix in line for line in tabs)
-    has_pinned = any(pinned_suffix in line for line in tabs)
-
-    for i, line in enumerate(tabs):
-        line = line.strip()
-        assert line.startswith('- ')
-        line = line[2:]  # remove "- " prefix
-
-        active = False
-        pinned = False
-
-        while line.endswith(active_suffix) or line.endswith(pinned_suffix):
-            if line.endswith(active_suffix):
-                # active
-                line = line[:-len(active_suffix)]
-                active = True
-            else:
-                # pinned
-                line = line[:-len(pinned_suffix)]
-                pinned = True
-
-        session_tab = session['windows'][0]['tabs'][i]
-        current_page = session_tab['history'][-1]
-        assert current_page['url'] == quteproc.path_to_url(line)
-        if active:
-            assert session_tab['active']
-        elif has_active:
-            assert 'active' not in session_tab
-
-        if pinned:
-            assert current_page['pinned']
-        elif has_pinned:
-            assert not current_page['pinned']
+    collapsed_suffix = ' (collapsed)'
+    # Don't check for states in the session if they aren't in the expected
+    # text.
+    has_active = any(active_suffix in line for line in expected_tabs)
+    has_pinned = any(pinned_suffix in line for line in expected_tabs)
+    has_collapsed = any(collapsed_suffix in line for line in expected_tabs)
+
+    def tab_to_str(tab, prefix="", collapsed=False):
+        """Convert a tab from a session file into a one line string."""
+        current = [
+            entry
+            for entry in tab["history"]
+            if entry.get("active")
+        ][0]
+        text = f"{prefix}- {current['url']}"
+        for suffix, state in {
+            active_suffix: tab.get("active") and has_active,
+            collapsed_suffix: collapsed and has_collapsed,
+            pinned_suffix: current["pinned"] and has_pinned,
+        }.items():
+            if state:
+                text += suffix
+        return text
+
+    def tree_to_str(node, tree_data, indentation=-1):
+        """Traverse a tree turning each node into an indented string."""
+        tree_node = node.get("treetab_node_data")
+        if tree_node:  # root node doesn't have treetab_node_data
+            yield tab_to_str(
+                node,
+                prefix="  " * indentation,
+                collapsed=tree_node["collapsed"],
+            )
+        else:
+            tree_node = node
+
+        for uid in tree_node["children"]:
+            yield from tree_to_str(tree_data[uid], tree_data, indentation + 1)
+
+    is_tree_tab_window = "treetab_root" in window
```

Comment:
I feel at some point we will be better off encapsulating treetab logic somewhere (I believe we already talked about this to some extent at least). Right now one has to understand treetab internals to ask for certain properties (e.g. if a window is a treetab window).

At this stage of development I am more than ok with domain logic floating around, though. Just leaving my thought here as a reminder for later.